lms/static/scripts/frontend_apps/components/test/SubmitGradeForm-test.js

```diff
@@ -293,6 +301,70 @@ describe('SubmitGradeForm', () => {
     });
   });
 
+  context('when page is about to be unloaded', () => {
+    let listener;
+
+    beforeEach(() => {
+      sinon
+        .stub(window, 'addEventListener')
+        .callsFake((eventName, callback) => {
+          listener = callback;
+        });
+      sinon.stub(window, 'removeEventListener');
+    });
+
+    afterEach(() => {
+      window.addEventListener.restore();
+      window.removeEventListener.restore();
+    });
```

Comment:
I didn't want to mock the listener, and instead my plan was to dispatch the actual event `window.dispatchEvent(new Event('beforeunload`))`, however, this causes karma to error and skip the test:

```
Chrome Headless 116.0.5845.110 (Linux x86_64) ERROR
  Some of your tests did a full page reload!
```

![image](https://github.com/hypothesis/lms/assets/2719332/8f0ea6e2-521a-4487-ab03-8a36f2147364)

That's why I have used this approach instead.

Reply:
I'm googling this though. I have found a couple threads of people suggesting possible workarounds, but either they seem to just want to test if the event listener is added, or their approach does not work for me.

Reply:
If you could extract the logic out into a hook:

```js
useWarnOnPageUnload(formIsDirty)
```

Where `formIsDirty` could either be a boolean state indicating whether there are unsaved changes, or a callback that can be invoked to check this, then you could isolate this ugliness to the hook's tests. The tests for this component would then just need to check that the correct dirty state value was passed.

Reply:
That's not a terrible idea. Let me give that a thought.